tt_metal/tt_stl/any_range.hpp

```diff
@@ -0,0 +1,887 @@
+};
+
+template <class TAnyIterator, class = iterator_category_t<TAnyIterator>>
+class IAnyIteratorAdaptor;
+
+template <class TAnyIterator>
+class IAnyIteratorAdaptor<TAnyIterator, std::input_iterator_tag> : public ICopyConstructible {
+   public:
+    [[nodiscard]] virtual bool operator==(const IAnyIteratorAdaptor &other) const = 0;
+    [[nodiscard]] virtual typename TAnyIterator::reference operator*() const = 0;
+    [[nodiscard]] virtual typename TAnyIterator::pointer operator->() const = 0;
+    virtual void operator++() = 0;
+};
+
+template <class TAnyIterator>
+class IAnyIteratorAdaptor<TAnyIterator, std::forward_iterator_tag>
+    : public IAnyIteratorAdaptor<TAnyIterator, std::input_iterator_tag> {
+   public:
+    using IAnyIteratorAdaptor<TAnyIterator, std::input_iterator_tag>::operator++;
+    [[nodiscard]] virtual TAnyIterator operator++(int) = 0;
+};
+
+template <class TAnyIterator>
+class IAnyIteratorAdaptor<TAnyIterator, std::bidirectional_iterator_tag>
+    : public IAnyIteratorAdaptor<TAnyIterator, std::forward_iterator_tag> {
+   public:
+    virtual void operator--() = 0;
+    [[nodiscard]] virtual TAnyIterator operator--(int) = 0;
+};
+
+template <class TAnyIterator>
+class IAnyIteratorAdaptor<TAnyIterator, std::random_access_iterator_tag>
+    : public IAnyIteratorAdaptor<TAnyIterator, std::bidirectional_iterator_tag> {
+   public:
+    virtual void operator+=(typename TAnyIterator::difference_type) = 0;
+    [[nodiscard]] virtual TAnyIterator operator+(typename TAnyIterator::difference_type) const = 0;
+    [[nodiscard]] virtual typename TAnyIterator::difference_type operator-(const IAnyIteratorAdaptor &) const = 0;
+    [[nodiscard]] virtual typename TAnyIterator::reference operator[](typename TAnyIterator::difference_type) const = 0;
+};
+
+template <class TAnyIterator, class TIterator, class TIteratorTag = iterator_category_t<TAnyIterator>>
+class AnyIteratorAdaptor;
+
+template <class TAnyIterator, class TIterator>
+class FinalAnyIteratorAdaptor final : public AnyIteratorAdaptor<TAnyIterator, TIterator> {
+    using AnyIteratorAdaptor<TAnyIterator, TIterator>::AnyIteratorAdaptor;
+};
+
+template <class TAnyIterator, class TIterator>
+class AnyIteratorAdaptor<TAnyIterator, TIterator, std::input_iterator_tag> : public IAnyIteratorAdaptor<TAnyIterator> {
+    using final_iterator_adaptor_type = FinalAnyIteratorAdaptor<TAnyIterator, TIterator>;
+
+   protected:
+    TIterator iterator;
+
+   public:
+    inline AnyIteratorAdaptor(const AnyIteratorAdaptor &) = default;
+    inline AnyIteratorAdaptor(AnyIteratorAdaptor &&) noexcept = default;
+
+    template <class UIterator, class = std::enable_if_t<std::is_same_v<remove_cvref_t<UIterator>, TIterator>>>
+    AnyIteratorAdaptor(UIterator &&iterator) : iterator(static_cast<UIterator &&>(iterator)) {}
+
+    [[nodiscard]] const type_info *id() const override {
+        static const type_info unique{};
+        return &unique;
```

Comment:
Here you are relying on the fact, that you should have unique addresses for static type_info instance. This behavior is not actually guaranteed, and some compilers (typically with LTO enabled) can actually merge those together resulting in different types having the same id. Can we please use std::type_index instead?

Reply:
This is untrue: https://stackoverflow.com/a/7526009/1541563 I don't want to depend on RTTI when this is perfectly sufficient.

Reply:
It is true. At the link you provided please read the comment with the green checkmark.
At Snapchat I spent a few days trying to catch this bug, and I don't really want to do it again.
@ayerofieiev-tt You should remember this bug with `zoo::Function`, right?

Reply:
@patrickroberts I would be okay with this, if you have an explicit test for this behavior in the tests.

Reply:
The answer with the green check mark outlines a particular version of a particular compiler that was pre-C++11. The optimization was a compiler bug, not a lack of guarantee by the standard.

Reply:
This still hods true for the modern compilers. For example, MSVC 2017 with C++17 and LTO enabled will perform such an optimization. Even though we don't compile with MSVC, we do support different compiler, and I would be skeptical to assume that there is no other compiler which does or will do this.
Please either use a different method for getting a unique type id, or add an explicit test to check that it does work as you expect.